pandas/core/indexes/category.py

```diff
@@ -430,7 +434,8 @@ def reindex(self, target, method=None, level=None, limit=None,
                 result = Index(np.array(self), name=self.name)
                 new_target, indexer, _ = result._reindex_non_unique(
                     np.array(target))
-
+                # see GH 16819, indexer needs to be converted to correct type
```

Comment:
instead handle in ``_reindex_non_unique``, furthermore we alway use ``dtype=np.intp_`` for indexers (and NOT int64 as it will fail on some platforms).

Need a test for this